File: frontend/src/backend/postService.js
import { appFetch, fetchConfig } from "./appFetch";

export const getPosts = (state, onSuccess, onErrors) => {
	appFetch(
		`/posts/?page=${state.page}&size=${state.size}`,
		fetchConfig("GET"),
		onSuccess,
		onErrors,
	);
};

export const postPosts = (post, onSuccess, onErrors) => {
	appFetch(`/posts/`, fetchConfig("POST", post), onSuccess, onErrors);
};

export const deletePost = (postId, onSuccess, onErrors) => {
	appFetch(`/posts/${postId}`, fetchConfig("DELETE"), onSuccess, onErrors);
};

export const editPost = (postId, post, onSuccess, onErrors) => {
	appFetch(`/posts/${postId}`, fetchConfig("PUT", post), onSuccess, onErrors);
};

export const getNRates = (postId, onSuccess, onErrors) => {
	appFetch(
		`/posts/nrates/${postId}`,
		fetchConfig("GET"),
		(countedRates) => {
			onSuccess(countedRates);
		},
		onErrors,
	);
};

export const getDidUserVote = (postId, onSuccess, onErrors) => {
	appFetch(
		`/posts/userrates/${postId}`,
		fetchConfig("GET"),
		(vote) => {
			onSuccess(vote);
		},
		onErrors,
	);
};

export const ratePost = (postId, rate, onSuccess, onErrors) => {
	appFetch(
		`/posts/${postId}/like/${rate}`,
		fetchConfig("POST"),
		onSuccess,
		onErrors,
	);
};

export const stillValid = (postId) => {
	appFetch(
		`/posts/${postId}/stillvalid`,
		fetchConfig("PATCH"),
	);	
};

export const createComment = (postId, comment, onSuccess, onErrors) => {
	appFetch(
		`/posts/${postId}/comment`,
		fetchConfig("POST", comment),
		onSuccess,
		onErrors,
	);
};

export const replyComment = (
	postId,
	comment,
	parentId,
	onSuccess,
	onErrors,
) => {
	appFetch(
		`/posts/${postId}/comment/${parentId}/reply`,
		fetchConfig("POST", comment),
		onSuccess,
		onErrors,
	);
};

export const deletePostComment = (postId, commentId, onSuccess, onErrors) => {
	appFetch(
		`/posts/${postId}/comment/${commentId}`,
		fetchConfig("DELETE"),
		onSuccess,
		onErrors,
	);
};

export const getPostComments = (postId, onSuccess, onErrors) => {
	appFetch(
		`/posts/${postId}/comments`,
		fetchConfig("GET"),
		(comments) => {
			onSuccess(comments);
		},
		onErrors,
	);
};

export const followPost = (postId, onSuccess, onErrors) => {
	appFetch(
		`/posts/${postId}/follow`,
		fetchConfig("POST"),
		onSuccess,
		onErrors,
	);
};

export const getPostFollows = (postId, onSuccess, onErrors) => {
	appFetch(
		`/posts/${postId}/follows`,
		fetchConfig("GET"),
		onSuccess,
		onErrors,
	);
};

export const deleteFollow = (postId, followId, onSuccess, onErrors) => {
	appFetch(
		`/posts/${postId}/follow/${followId}`,
		fetchConfig("DELETE"),
		onSuccess,
		onErrors,
	);
};

File: src/main/java/es/udc/fi/dc/fd/model/services/PostService.java
package es.udc.fi.dc.fd.model.services;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.data.domain.Page;

import es.udc.fi.dc.fd.model.common.exceptions.InstanceNotFoundException;
import es.udc.fi.dc.fd.model.entities.Category;
import es.udc.fi.dc.fd.model.entities.Comment;
import es.udc.fi.dc.fd.model.entities.Follow;
import es.udc.fi.dc.fd.model.entities.Post;
import es.udc.fi.dc.fd.model.entities.Rate;
import es.udc.fi.dc.fd.model.services.exceptions.IncorrectPriceException;
import es.udc.fi.dc.fd.rest.dtos.RatesCountDto;

public interface PostService {

        Post createPost(Long userId, String title, String description, String url, Double price, List<String> images,
                        Long category, Post.Type type, String couponCode, LocalDateTime expirationDate)
                        throws InstanceNotFoundException, IncorrectPriceException;

        Page<Post> getLatestPosts(int page, int size);

        void deletePost(Long postId, Long userId) throws InstanceNotFoundException;

        void updatePost(Long postId, Long userId, String title, String description, String url, Double price,
                        List<String> images, boolean isActive)
                        throws InstanceNotFoundException, IncorrectPriceException;

        void ratePost(Long postId, Long userId, boolean rate) throws InstanceNotFoundException;

        RatesCountDto getPostNRates(Long postId) throws InstanceNotFoundException;

        Rate getUserRate(Long userId, Long postId) throws InstanceNotFoundException;

        List<Category> getCategories();

        void stillValid(Long postId) throws InstanceNotFoundException;

        List<Rate> getPostRates(Long postId) throws InstanceNotFoundException;

        List<Comment> getComments();

        List<Comment> getPostComments(Long postId) throws InstanceNotFoundException;

        Comment createComment(Long userId, Long postId, String text) throws InstanceNotFoundException;

        Comment replyComment(Long userId, Long postId, Long parentCommentId, String text)
                        throws InstanceNotFoundException;

        void updateComment(Long userId, Long commentId, String text) throws InstanceNotFoundException;

        void deleteComment(Long userId, Long commentId) throws InstanceNotFoundException;

        Comment getComment(Long commentId) throws InstanceNotFoundException;
        
        Follow createFollow(Long userId, Long postId) throws InstanceNotFoundException;
        
        List<Follow> getPostFollows(Long postId) throws InstanceNotFoundException;
        
        void deleteFollow(Long followId) throws InstanceNotFoundException;

}
